### PR TITLE
Ac 684 combine receipts

### PR DESCRIPTION
**Pull request by PaulSnow**
FIX: FEAT:
Fixed the receipts.  Added tests that prove both that the receipt proves a hash (drives to a particular Merkle Dag Root, and that the Merkle Dag Root matches the Merkle Dag Root generated by the Merkle Tree for the Anchor point).
Added a combine function to combine two receipts.  This actually went in with other related pull requests, but the tests are enabled here.
Added a MerkleManager.GetIntermediate() call that returns the two hashes that were combined to produce a particular hash at an index and at a given height.

### CODE REVIEW

File: smt/managed/merklemanager.go [hunk 1]
```diff
@@ -292,14 +300,33 @@ func (m *MerkleManager) Get(element int64) (Hash, error) {
...
-// GetNext
-// Get the next hash to be added to a state at this height
-func (m *MerkleManager) GetNext(element int64) (hash Hash) {
-	data, err := m.Manager.Get(m.key.Append("NextElement", element))
-	if err != nil {
-		return nil
+// GetIntermediate
+// Return the last two hashes that were combined to create the local
+// Merkle Root at the given index.  The element provided must be an
+// even power of two >= 2.
```
Comment (firelizzard18) [inline smt/managed/merklemanager.go:306]: > The element provided must be an even power of two >= 2.

I tried enforcing this, but adding any of the three if-statements below causes the tests to fail.

```go
if element < 4 {
	return nil, nil, fmt.Errorf("element is too small")
}
if bits.OnesCount64(uint64(element)) != 1 {
	return nil, nil, fmt.Errorf("element is not a power of 2")
}
if l := bits.Len64(uint64(element)) - 1; l%2 != 0 {
	return nil, nil, fmt.Errorf("element is not an even power of 2")
}
```

File: smt/managed/merklestate.go [hunk 2]
```diff
@@ -79,6 +79,25 @@ func (m *MerkleState) PadPending() {
...
+// Trim
+// Remove any trailing nils from Pending hashes
+func (m *MerkleState) Trim() {
+	for len(m.Pending) > 0 && m.Pending[len(m.Pending)-1] == nil {
+		m.Pending = m.Pending[:len(m.Pending)-1]
+	}
+}
+
+// Pad
+// Add a til to the end of the Pending list if one isn't there.
```
Comment (firelizzard18) [inline smt/managed/merklestate.go:91]:
```suggestion
// Add a nil to the end of the Pending list if one isn't there.
```

File: smt/managed/receipt.go [hunk 3]
```diff
@@ -222,66 +50,176 @@ func (r *Receipt) String() string {
 	return b.String()
 }
 
+// Validate
+// Take a receipt and validate that the element hash progresses to the
+// Merkle Dag Root hash (MDRoot) in the receipt
+func (r Receipt) Validate() bool {
```
Comment (firelizzard18) [inline smt/managed/receipt.go:56]:
```suggestion
func (r *Receipt) Validate() bool {
```

File: smt/managed/receipt.go [hunk 4]
```diff
@@ -222,66 +50,176 @@ func (r *Receipt) String() string {
...
+			MDRoot = hash.Combine(Sha256, MDRoot)
+		}
+	}
+	// In the end, MDRoot should be the same hash the receipt expects.
+	return MDRoot.Equal(r.MDRoot)
+}
+
+// Copy
+// Make a copy of this receipt
+func (r Receipt) Copy() *Receipt {
```
Comment (firelizzard18) [inline smt/managed/receipt.go:76]:
```suggestion
func (r *Receipt) Copy() *Receipt {
```

File: smt/managed/receipt.go [hunk 5]
```diff
@@ -222,66 +50,176 @@ func (r *Receipt) String() string {
...
+func (r *Receipt) Combine(rm *Receipt) (*Receipt, error) {
+	if !bytes.Equal(r.MDRoot, rm.Element) {
+		return nil, fmt.Errorf("receipts cannot be combined. "+
+			"anchor %x doesn't match root merkle tree %x", r.Anchor, rm.Element)
+	}
+	nr := r.Copy()
+	nr.MDRoot = rm.MDRoot
+	for _, n := range rm.Nodes {
+		nr.Nodes = append(nr.Nodes, n)
+	}
```
Comment (firelizzard18) [inline smt/managed/receipt.go:103]:
```suggestion
	nr.Nodes = append(nr.Nodes, rm.Nodes...)
```

File: smt/pmt/bpt_test.go [hunk 6]
```diff
@@ -402,8 +402,11 @@ func walk(bpt *BPT, node *Node) {
 	}
 }
 
+// Just gives data about distributions in the BPT
 func TestBPTByteSizes(t *testing.T) {
-	t.Skip("Just gives data about distributions in the BPT.")
+	if true {
+		return
+	}
```
Comment (firelizzard18) [inline smt/pmt/bpt_test.go:409]: ?

**Review by firelizzard18 — CHANGES_REQUESTED** [review]

File: smt/managed/merklemanager.go (hunk 1 above)
Comment (PaulSnow) [reply smt/managed/merklemanager.go:306]: The comment is wrong.  You can create a Merkle Root at any index.  I'll go fix it.

File: smt/managed/receipt_test.go [hunk 7]
```diff
@@ -73,49 +75,57 @@ func TestReceipt(t *testing.T) {
...
 					t.Fatal("Should not be able to generate a receipt")
 				}
 			} else {
+				require.Nilf(t, err, "Failed to get a receipt: %v", err)
 				if r == nil {
 					t.Fatal("Failed to generate receipt", i, j)
 				}
 				if !r.Validate() {
-					t.Fatal("Receipt fails for element ", i, " anchor ", j)
+					fmt.Printf("Receipt fails for element/anchor [%d %d]\n", i, int64(j))
```
[the remark this answers is not in the document]
Comment (PaulSnow) [reply smt/managed/receipt_test.go:117]: Should throw the error

[the remark this answers is not in the document]
Comment (firelizzard18) [reply smt/managed/receipt_test.go:117]: Throw as in `panic(err)`? If the test should stop immediately, `t.Fatal` will do that without generating a full stack trace.

File: smt/managed/receipt.go (hunk 5 above)
Comment (PaulSnow) [reply smt/managed/receipt.go]: Resolved.  Should have been copying the nodes.

**Review by firelizzard18 — APPROVED** [review]

**Pull request merged**